### PR TITLE
Dotnet template

### PR DESCRIPTION
**Pull request by RobotOptimist**
#14 

A dotnet new template in a template folder. Also included is a nuspec file.
I was able to test the installation of the template successfully. 
I was unable to test the use of the template because the nuget package has not been published yet
I assumed the version would be 0.1.31-alpha, but this will need updated.

I can make this work using your feedzio by adding a nuget.config and changing the version to match that prerelease - let me know if you need those changes.

### CODE REVIEW

File: template/FeatherHttpTemplate.nuspec [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package>
+    <metadata>
+        <id>FeatherHttpTemplate</id>
```
Comment (davidfowl) [inline template/FeatherHttpTemplate.nuspec]: FeatherHttp.Templates

**Comment by davidfowl** [comment]
> I can make this work using your feedzio by adding a nuget.config and changing the version to match that prerelease - let me know if you need those changes.

Yes this would be great. Can we also integrate it into the github actions script.

File: template/FeatherHttpTemplate.nuspec (hunk 1 above)
Comment (RobotOptimist) [reply template/FeatherHttpTemplate.nuspec]: No problem, would you like all instances of template to be pluralized in this pr? I thought you only needed name in template.json that way, sorry for the confusion 😅

**Comment by davidfowl** [comment]
> I assumed the version would be 0.1.31-alpha, but this will need updated.

Yes, I also broke a bunch of APIs

**Comment by RobotOptimist** [comment]
I've added the nuget config and changed the version to match what's in feedz.io but I haven't figured out how to make these things update automatically in the ci.yml yet. Looks like I can update these files using powershell and check them in using actions - but I don't know how to discover the new version number while the actions are running. 
I looked at the Carter repo to see if I can get some insight, but it doesn't appear they're doing that with github actions. I'm continuing to investigate the problem to figure it out - but I'm held up for now

**Comment by davidfowl** [comment]
There’s also code changes that need to be made

**Comment by RobotOptimist** [comment]
yikes! So that's what you meant by API changes, I've pulled that in - still reading up on the automatic version incrementing -trying to see if Nerdbank has documentation that can help for this template

**Comment by RobotOptimist** [comment]
Okay, I've finally got something working - it's in a branch ci-changes since it required a lot of testing using github actions that I can't simulate on my local machine. For that reason, the commit history on that branch is long. I want to merge it to this branch and squash the commits - let me know if that's ok, I'm not sure what your policy is on losing some commit history like that.

It does commit against the branch being worked on - but you may at some point in the future want to only allow it to change if it's working against master, or a branch you set aside for the template nuspec. 

I've tested the ci build against my personal feedz.io and validated that the workflow is not disrupted by my changes. 

I further intend to add the same conditions to the template nuspec version update steps that you apply to your nuget pushes.

In short, all I need to proceed is to know your preference on squashed commits. 
This is my first PR to an open source library like this and I'm really excited about it, but also open to feedback. If you find the time, will you please help me by  pointing me toward some guiding standards so I can improve at open source contributions?

**Comment by davidfowl** [comment]
Squashing is fine! That’s great. Thanks for that level of testing! 

Side question: is it possible to build a template with a csproj?

**Comment by RobotOptimist** [comment]
Oh, you mean test the build of the template project in the ci? There already is a csproj file included in this PR at template/content/featherhttp-template.csproj
I'll work on some additional CI changes so we can verify that the template project builds correctly

**Comment by davidfowl** [comment]
Ah no thats fine. Is this good to go? Are you going to merge?

**Comment by RobotOptimist** [comment]
Good to go, let me quickly go over the lessons learned from doing this in case it's helpful later, also let me know if you want these items documented elsewhere.
Lessons learned:

1. Nerdbank git versioning will cause errors if you do dotnet builds on subsequent jobs due to errors on reading the version.json file. It has something to do with clone depth, I found a solution from an issue - but it was appveyor specific and could not be used to resolve the github actions error.
2. If you install the nbgv tool prior to pushing the packages to feedz.io then be aware that it will be packaged and sent along with everything else. It is not downloaded when someone pulls down the package though, but might be additional overhead that you don't want to deal with.
3. The versioning tool sometimes has a variance of a few minor versions. It uses commit history and sorcery to determine what the version will be. 

Some stuff to be aware of:

1. Right now, I'm using the nbgv tool get-version command to find the version, this might cause problems later if there is a precision problem with versioning between the nuget package and the template csproj/template nuspec. (this has not mattered too much since nuget will resolve the issue by getting the closest version)
2. The template nuspec (tries to have) will have the same version as the project nuget package.
3. To test .ps1 file locally you'll need to install the nbgv tool globally, the command is documented in the  .ps1 file
4. If you push to master, this is set up to commit against that branch and you'll need to pull down again after you do a push.

Recommendation: If the potential nbgv imprecision looks to be causing issues then we'll need to read the version out of the built assemblies somehow.

It was fun doing this - I'm looking forward to getting the docs site setup next.

**Comment by RobotOptimist** [comment]
@davidfowl  It's good you waited to merge this.
I've removed the nbgv dependency completely, reading the version from the dll with powershell. This ensures a consistent version across all files.
Removing the nbgv tool dependency removes most of the areas of concern outlined in my previous comment. You will still need to pull after pushing to master. 
This does mean that the github version is being saved to a file called NUGET_VERSION on the github build server so that it's accessible in subsequent github action steps.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -40,4 +44,39 @@ jobs:
...
+      if: github.event_name == 'push' && github.ref == 'refs/heads/master'
+      uses: ad-m/github-push-action@master
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        branch: ${{ github.ref }}
+
+    - name: test build template
+      if: github.event_name == 'push' && github.ref == 'refs/heads/master'
+      run: dotnet build ./template/content/featherhttp-template.csproj
+
```
Comment (davidfowl) [inline .github/workflows/ci.yaml]: nit: lots of extra space here.

**Comment by davidfowl** [comment]
Here's what I'm wondering. Instead of using a nuspec file, would it be possible to produce a template package from a csproj? The flow would look like this:

- Replace the [template](https://github.com/davidfowl/FeatherHttp/blob/02d7f0adb92ed911521ef40838a55f35c3295702/template/FeatherHttpTemplate.nuspec) with a csproj that includes the gitversion package (that should be automatic since its auto included everywhere). You should be able to set all of the same properties.
- As part of the build process for that csproj, we should be generating [this csproj](https://github.com/davidfowl/FeatherHttp/blob/02d7f0adb92ed911521ef40838a55f35c3295702/template/content/featherhttp-template.csproj#L1) with the correct version. That should be easy to do since the git version information should be available during the build.

Removes powershell and windows specific logic.

**Comment by RobotOptimist** [comment]
I'm not sure if I can do it - but I'm willing to try. It's a departure from the Carter project too. Certainly understand wanting to remove the powershell dependency though.
One more thing I need to know - with moving away from the nuspec - how will users be installing the template?

**Comment by davidfowl** [comment]
You can generate a package without a nuspec (like the main feather project)

**Comment by RobotOptimist** [comment]
@davidfowl I want to give you an update on this since it's taking so long. I've managed to accomplish the task *mostly*, there is an issue using the NerdBank msbuild variable to set the version in the csproj files. I followed what documentation there was on that, but ultimately resorted back to powershell - so there are a couple lines of powershell still. I apologize - but looking at other github workflow files from other projects, I'm not convinced it can be avoided.
So, that's all working great and I can upload the template project to feedz.io and that works fine. 

While testing the template project to make sure it could be used as expected I ran against this bug:
https://github.com/dotnet/templating/issues/1423
So now I'm trying to find a way to override the version but only for the Feather Template project so that it can be consumed by users. I haven't opened an issue with NerdBank yet, but I've begun to engage them in their gitter: https://gitter.im/AArnott/Nerdbank.GitVersioning

You'll note that we're not the only ones having issues with NerdBank in github actions. I think as a side effect of this effort I'll do a PR to improve their documentation on this. Just little gotchas, but they have a great tool and I understand it much better than I did before.

Anyway, I'm deeply grateful for the opportunity to contribute anything meaningful to this project and I apologize for the delay in accomplishing this task.

**Comment by davidfowl** [comment]
I see, I guess we can live with it for now then.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -40,4 +44,39 @@ jobs:
...
+      with:
+        ref: ${{ github.ref }} 
+
+    - name: nuget version update
+      if: github.event_name == 'push' && github.ref == 'refs/heads/master'
+      run: | 
+        powershell ./update-nuget-references.ps1
+        git status        
+
+    - name: Commit files
```
Comment (davidfowl) [inline .github/workflows/ci.yaml]: What's this for?

Comment (RobotOptimist) [reply .github/workflows/ci.yaml]: This existed because I misunderstood the requirements. This has now been changed

**Comment by RobotOptimist** [comment]
okay - I've got this working
1. I've rewritten the process for creating the template to do it the way you suggested. 
2. I've refactored my csproj names and directories to be consistent with project naming conventions
3. I've added a version.json and modified the ci.yml to use the latest and greatest from actions/checkout and to perform stable releases of the template project so it can be used
    a. you'll notice this does a checkout of a branch (in this case master) and fetches commit history - this is necessary to use the nbgv features the way they were intended to be used
4. If this is not pushed to nuget.org then users will need to add the feedz.io source in order to use this with `dotnet new -i FeatherHttp.Templates` afterwards they'll be able to create the hello world project by typing `dotnet new featherhttp`

I've learned a ton about github actions, nbgv, and template projects as a result of this - it's been a pretty great experience

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (RobotOptimist) [reply .github/workflows/ci.yaml]: So specifically - what I thought you wanted was to get the version info and update the csproj files in mater with that version info. 
But after looking at other implementations at other projects and reading and thinking more about what you may want - I thought you probably just want the uploaded nuget file csproj to have the correct version and there is no reason to do a commit from the ci build. 
Tbh, I had misunderstood the requirement for this issue and that's why the PR has changed so much. I think it's pretty close in it's current state

**Comment by RobotOptimist** [comment]
Hey @davidfowl now that I've got a better understanding of what's needed, this branch is a bit of a mess - what do you think of me closing the PR and setting up a fresh branch with a cleaner commit  history that arrives at the goal a bit more directly?

**Comment by davidfowl** [comment]
Sounds good to me. Do you want me to close it? You can as well

**Comment by RobotOptimist** [comment]
I'll take care of it, thanks David
